cmd/bisync/deltas.go

```diff
@@ -0,0 +1,310 @@
+// Package bisync implements bisync
+// Copyright (c) 2017-2020 Chris Nelson
+package bisync
+
+import (
+	"context"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/rclone/rclone/cmd/bisync/bilib"
+	"github.com/rclone/rclone/fs"
+	"github.com/rclone/rclone/fs/operations"
+)
+
+// delta
+type delta uint8
+
+const (
+	deltaZero delta = 0
+	deltaNew  delta = 1 << iota
+	deltaNewer
+	deltaOlder
+	deltaSize
+	deltaHash
+	deltaDeleted
+)
+
+const (
+	deltaModified delta = deltaNewer | deltaOlder | deltaSize | deltaHash | deltaDeleted
+	deltaOther    delta = deltaNew | deltaNewer | deltaOlder
+)
+
+func (d delta) is(cond delta) bool {
+	return d&cond != 0
+}
+
+// deltaSet
+type deltaSet struct {
+	deltas     map[string]delta
+	opt        *Options
+	fs         fs.Fs  // base filesystem
+	msg        string // filesystem name for logging
+	oldCount   int    // original number of files (for "excess deletes" check)
+	deleted    int    // number of deleted files (for "excess deletes" check)
+	foundSame  bool   // true if found at least one unchanged file
+	checkFiles bilib.Names
+}
+
+func (ds *deltaSet) empty() bool {
+	return len(ds.deltas) == 0
+}
+
+func (ds *deltaSet) sort() (sorted []string) {
+	if ds.empty() {
+		return
+	}
+	sorted = make([]string, 0, len(ds.deltas))
+	for file := range ds.deltas {
+		sorted = append(sorted, file)
+	}
+	sort.Strings(sorted)
+	return
+}
+
+func (ds *deltaSet) printStats() {
+	if ds.empty() {
+		return
+	}
+	nAll := len(ds.deltas)
+	nNew := 0
+	nNewer := 0
+	nOlder := 0
+	nDeleted := 0
+	for _, d := range ds.deltas {
+		if d.is(deltaNew) {
+			nNew++
+		}
+		if d.is(deltaNewer) {
+			nNewer++
+		}
+		if d.is(deltaOlder) {
+			nOlder++
+		}
+		if d.is(deltaDeleted) {
+			nDeleted++
+		}
+	}
+	fs.Infof(nil, "%s: %4d changes: %4d new, %4d newer, %4d older, %4d deleted",
+		ds.msg, nAll, nNew, nNewer, nOlder, nDeleted)
+}
+
+// findDeltas
+func (b *bisyncRun) findDeltas(fctx context.Context, f fs.Fs, oldListing, newListing, msg string) (ds *deltaSet, err error) {
+	var old, now *fileList
+
+	old, err = b.loadListing(oldListing)
+	if err != nil {
+		fs.Errorf(nil, "Failed loading prior %s listing: %s", msg, oldListing)
+		b.abort = true
+		return
+	}
+	if err = b.checkListing(old, oldListing, "prior "+msg); err != nil {
+		return
+	}
+
+	now, err = b.makeListing(fctx, f, newListing)
+	if err == nil {
+		err = b.checkListing(now, newListing, "current "+msg)
+	}
+	if err != nil {
+		return
+	}
+
+	ds = &deltaSet{
+		deltas:     map[string]delta{},
+		fs:         f,
+		msg:        msg,
+		oldCount:   len(old.list),
+		opt:        b.opt,
+		checkFiles: bilib.Names{},
+	}
+
+	for _, file := range old.list {
+		d := deltaZero
+		if !now.has(file) {
+			b.indent(msg, file, "File was deleted")
+			ds.deleted++
+			d |= deltaDeleted
+		} else {
+			if old.getTime(file) != now.getTime(file) {
+				if old.beforeOther(now, file) {
```

Comment:
@ncw

This approach is simplistic based on fact that we always compare a filesystem with itself. Compare this e.g. with [equal()](https://github.com/rclone/rclone/blob/v1.55.1/fs/operations/operations.go#L167) or even with proposal #4810.

Reply:
However, @Jwink3101 is right. (In)equality and before/after relations in this piece should be rethought with backend's time window in mind.

Reply:
Something like [fstest.CheckTimeEqualWithPrecision] (https://github.com/rclone/rclone/blob/v1.55.1/fstest/fstest.go#L116) for equaliy but extended to the `>` and `<` relations. I'd put it in separate functions and add unit tests that ensure it's [anticommutative](https://en.wikipedia.org/wiki/Anticommutative_property) etc.

Reply:
I have added a related _optional item_